File: C++/프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include <queue>
#include <algorithm>

using namespace std;

int n, m;
int dx[4] = {0, 0, 1, -1};
int dy[4] = {1, -1, 0, 0};

int visited[101][101];

void bfs (int sx, int sy, const vector<vector<int>>& map) {
    queue<pair<int, int>> q;
    visited[sx][sy] = 1;
    q.push({sx, sy});
    
    while (!q.empty()) {
        int cx = q.front().first;
        int cy = q.front().second;
        q.pop();
        
        if (cx == n - 1 && cy == m - 1) {
            return;
        }
        
        for (int i = 0; i < 4; i++) {
            int nx = cx + dx[i];
            int ny = cy + dy[i];
            
            if (nx >= 0 && ny >= 0 && nx < n && ny < m) {
                if (!visited[nx][ny] && map[nx][ny] != 0) {
                    q.push({nx, ny});
                    visited[nx][ny] = visited[cx][cy] + 1;
                }
            }
        }
    }
}

int solution(vector<vector<int> > maps)
{
    int answer = 0;
    n = maps.size();
    m = maps[0].size();
    
    bfs (0, 0, maps);
    
    answer = visited[n - 1][m - 1];
    if (answer == 0) {
        return -1;
    }
    return answer;
}

File: Java/프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    
    /*
        최단거리니까 BFS일 거야.
        상하좌우 갈 수 있어
    */
    static int[] dx = {0, 0, 1, -1};
    static int[] dy = {1, -1, 0, 0};
    static int[][] visited;
    
    static void bfs (int sx, int sy, int[][] maps) {
        Queue<int[]> q = new ArrayDeque<>();
        q.offer(new int[]{sx, sy});
        visited[sx][sy] = 1;
        
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            
            if (cur[0] == maps.length - 1 && cur[1] == maps[0].length - 1) {
                return;
            }
            
            for (int i = 0; i < 4; i++) {
                int nx = cur[0] + dx[i];
                int ny = cur[1] + dy[i];
                
                if (nx >= 0 && nx < maps.length && ny >= 0 && ny < maps[0].length) {
                    if (maps[nx][ny] == 1 && visited[nx][ny] == 0) {
                        visited[nx][ny] = visited[cur[0]][cur[1]] + 1;
                        q.offer(new int[]{nx, ny});
                    }
                }
            }
        }
    }
    
    public int solution(int[][] maps) {
        int answer = 0;
        visited = new int[maps.length][maps[0].length];
        
        bfs(0, 0, maps);
        answer = visited[maps.length - 1][maps[0].length - 1];
        if (answer == 0) {
            return -1;
        }
        return answer;
    }
}